File: src/Chess/Piece.java
package Chess;

import java.util.LinkedList;

public abstract class Piece {

    public enum Colour {
        NONE  (' '),
        WHITE ('w'),
        BLACK ('b');

        public final char symbol;

        Colour(char symbol) {
            this.symbol = symbol;
        }
    };

    public final Colour colour;
    public final char symbol;
    public final int value;
    
    // public static final Piece NO_PIECE = new NullPiece();

    // public static final Piece WHITE_PAWN = new Pawn(Colour.WHITE);
    // public static final Piece BLACK_PAWN = new Pawn(Colour.BLACK);

    // TODO static constants for each piece (maybe in Piece Catalogue)

    public Piece(Colour colour, char symbol, int value) {
        this.colour = colour;
        this.symbol = symbol;
        this.value = value;
    }

    public String getSymbol() {
        return "" + colour.symbol + this.symbol;
    }

    public boolean equals(Piece piece) {
        return (this.colour == piece.colour && this.symbol == piece.symbol);
    }

    public abstract LinkedList<Position> getLegalMoves(Board board, Position position);
}

File: chess/Piece.cpp

#include "Piece.hpp"

const Piece Piece::NO_PIECE(NO_COLOUR, NO_TYPE);

Piece::Piece() {}

Piece::Piece(Colour colour, PieceType type) :
    colour(colour), type(type)
{}

bool Piece::equals(const Piece& piece) const {
    return (
        this->colour == piece.colour &&
        this->type == piece.type
    );
}

std::string Piece::getSymbol() const {
    char symbol[3] = {colour, type, 0};
    return std::string(symbol);
}